### PR TITLE
7312 export with map name

### LINKED ISSUE #7312: Use the map name for the map export file
cc @urbanphes @gfiorav

### PR DESCRIPTION
**Pull request by gfiorav**
Fixes #7312 

Please CR @juanignaciosl 

The format is:

`<map name> (on <YYYY-MM-DD> at <HH.MM.SS>).carto`

example

`Sequías en España (on 2016-05-31 at 17.34.55).carto`

All special characters are preserved on exported filename and imported map name.

### CODE REVIEW

**Comment by juanignaciosl** [comment]
1. Has the format been agreed with @urbanphes ? He suggested the visualization name and there's been no further discussion at #7312.
2. Have you checked the behaviour with other characters? For example, "1Vïs / Map \ Whaterver ·/.. ..." is a valid visualization name and I hardly believe it to be a valid name file.
3. I miss a test.

**Comment by juanignaciosl** [comment]
4. And please check both in local and staging, since using a directory vs using S3 has different naming limitations.

**Comment by gfiorav** [comment]
Works in staging and local.

  

Seems a valid name according to different "valid characters for filenames"
charts on line. Only "weird" character i'm not 100% sure of is `(`,')`.

  

I'll check with Juanma and add a test tomorrow

**Comment by gfiorav** [comment]
Hey, @juanignaciosl, find a reviewed version.

I know sanitise filenames to avoid forbidden characters (I can't only allow ASCII characters as our customers use heavily non-ASCII languages (Japanese, Chinese, etc).

Juanma has approved the name format.

File: spec/lib/carto/file_system/sanitize_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      describe '#sanitize_identifier' do
+        it 'should allow for non ascii characters' do
+          identifier = 'MæđrÏD, ĘŠPªÑÅ'
+
+          Carto::FileSystem::Sanitize.sanitize_identifier(identifier).should eq identifier
+        end
+
+        it 'should sanitize disallowed characters' do
+          identifier = "M\x00æđ/rÏD,\\:* : ! ? \" ' >Ę<ŠP|ªÑÅ"
+          sanitized_identifier = 'M_æđ_rÏD,___ _ ! _ _ \' _Ę_ŠP_ªÑÅ'
```
Comment (juanignaciosl) [inline spec/lib/carto/file_system/sanitize_spec.rb]: 1.- Do we really want to accept this kind of characters for filenames? :-\
2.- Add a test incluing `/`, please. `"1Vïs / Map \ Whaterver ·/.. ..." `, for example.

Ping me after tests pass and branch is mergeable, please.

Comment (gfiorav) [inline spec/lib/carto/file_system/sanitize_spec.rb]: 1. If we only allow ASCII, Koreans will get filenames like `___________.carto`. Files with these characters in their name are not only common, but obligatory for some languages. File systems deal with them just fine.

2. This test has both `/` and `\`.

Comment (juanignaciosl) [inline spec/lib/carto/file_system/sanitize_spec.rb]: Up to you then (I had overlooked the slashes, which was my main concern).

Comment (gfiorav) [inline spec/lib/carto/file_system/sanitize_spec.rb]: I just did a staging test using a map name `M\x00æđ/rÏD,\\:* : ! ? \" ' >Ę<ŠP|ªÑÅ`.

Worked out perfect, got: `M_x00æđ_rÏD,____ _ ! _ __ ' _Ę_ŠP_ªÑÅ (on 2016-05-17 at 13.20.40).zip`

Seems pretty bomb proof 🎌

**Comment by gfiorav** [comment]
@juanignaciosl I just noticed that I merged some stuff wrong. Check again please, there are some "allow_spaces" params in the uploader that I'm not sure you approve of.

File: lib/carto/file_system/sanitize.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module Carto
+  module FileSystem
+    module Sanitize
+      # Sanitize: NUL, slash, backslash, colon, asterisk, question, quote, less, greater, pipe
+      DISALLOWED_CHARACATERS = /[\x00\/\\:\*\?\"<>\|]/
```
Comment (juanignaciosl) [inline lib/carto/file_system/sanitize.rb]: Typo.

**Comment by juanignaciosl** [comment]
If it works, LGTM. I'm not a fan of spaces but my biggest concerns were the slashes. If you've checked that both with local filesystem and S3 the full process works, ✨

**Comment by gfiorav** [comment]
Thanks, I'll check again in staging for good measure, but it's been thoroughly tested in both environments.

**Comment by gfiorav** [comment]
Retest this, please

**Pull request merged**